generate.py

```diff
@@ -55,23 +55,32 @@ def logits_to_probs(logits, temperature: float = 1.0, top_k: Optional[int] = Non
     return probs
 
 
-def sample(logits, temperature: float = 1.0, top_k: Optional[int] = None):
-    probs = logits_to_probs(logits[0, -1], temperature, top_k)
+def sample(logits, temperature: float = 1.0, top_k: Optional[int] = None, ignore_gist: bool = True):
+    next_logits = logits[0, -1]
+    if ignore_gist:
+        next_logits[-1] = 0
+    probs = logits_to_probs(next_logits, temperature, top_k)
     idx_next = multinomial_sample_one_no_sync(probs)
     return idx_next, probs
 
-
-def prefill(
-    model: Transformer, x: torch.Tensor, input_pos: torch.Tensor, **sampling_kwargs
-) -> torch.Tensor:
-    # input_pos: [B, S]
+def get_mask(input_ids: torch.Tensor, input_pos: torch.Tensor, gist_token_id: int):
     causal_mask = (
         torch.tril(torch.ones(len(input_pos), len(input_pos), dtype=torch.bool))
         .unsqueeze(0)
         .unsqueeze(0)
-        .to(x.device)
+        .to(input_ids.device)
     )
-    logits = model(x, input_pos, mask=causal_mask)
+    gist_token_positions = torch.stack(torch.where(input_ids == gist_token_id)).T
+    for position in gist_token_positions:
+        causal_mask[position[0], :, position[1] + 1:, :position[1]] = False
+    
+    return causal_mask
+
+def prefill(
+    model: Transformer, x: torch.Tensor, input_pos: torch.Tensor, **sampling_kwargs
+) -> torch.Tensor:
+    # input_pos: [B, S]
+    logits = model(x, input_pos, mask=get_mask(x, input_pos, 128256))
```

Comment:
The GIST token ID should not be hardcoded and `get_mask` should apply to non GIST models as well. It might be cleaner to just write a `prefill_gist` function to separate the logic